### ISSUE TITLE
api: skip AuthenticatedActionView authentication for safe HTTP methods

### ISSUE THREAD

**Issue by peterthomassen** [desec-io/desec-stack#494]
When GET'ing a confirmation link such as `https://desec.io/api/v1/v/activate-account/.../`, we send a redirect to `https://desec.io/confirm/activate-account/.../` so that the GUI appears, which then retries the initial URL via POST.

Currently, the view performs authentication via `AuthenticatedBasicUserActionAuthentication`, which fails if the verification code is invalid (e.g., expired). As a result, the redirect does not happen, and browsers display an ugly `400 Bad Request` message without any explanation. This is because the authenticator uses `serializer.is_valid(raise_exception=True)`, which raises an exception in this case, but the HTML renderer does not render the exception detail.

If we change the `authentication_classes` attribute into a dynamic property, such as
```
    @property
    def authentication_classes(self):
        return () if self.request.method in SAFE_METHODS else (auth.AuthenticatedBasicUserActionAuthentication,)
```
then this problem disappears, and users get a nice error message in the GUI.

@nils-wisiol Do you think there are any adverse implications in this approach?

**Comment by nils-wisiol** [comment]
A summary of the behavior of `AuthenticatedActionView` depending on HTTP method and Accept header:

| `AuthenticatedActionView` | GET | POST |
| --- | ------ | --------- |
| Accept: html | forward to `self.html_url` if any | perform action
| else | HTTP 406 Not Acceptable | perform action

Other HTTP methods are all 406 Not Acceptable.

The redirect is essentially performed by `redirect(self.html_url.format(**kwargs))`, where `self.html_url` is a constant and `kwargs` are the passed to the `get` method and provided by the request. Currently, the only kwarg is `code`, which contains the full auth code as provided by the request.

So, by not authenticating `GET` requests, we do not expose any information nor take we any action on the server side, so I do not see any problem.

However, I see that in `AuthenticatedActionView`, `perform_authentication` is overwritten to do nothing, and `super().perform_authentication(request)` is called explicitly in `post`, but not in `get`. I'm not sure why requests are authenticated when using `GET`, but I can confirm they are by running `http GET https://desec.io/api/v1/v/activate-account/foobar/`.

**Comment by peterthomassen** [comment]
Regarding the last aspect of your comment: We introduced the lazy authentication in a753f2222e4e3f79b1b4cd0b581eba19801c9e52, but later introduced rate limiting in 96c85bfb59ed11d402191fd3cd6f3fff6f07d095. The rate limiting stuff calculates a user-specific cache key, so [it accesses `request.user`](https://github.com/encode/django-rest-framework/blob/d635bc9c71b6cf84d137a68610ae2e628f8b62b3/rest_framework/throttling.py#L242), and that triggers authentication.

The fix I proposed in the initial report should also remedy this. We can then remove the `perform_authentication()` override, as authentication for safe methods will be no-op anyways.
